probably/pgcl/syntax.py

```diff
@@ -125,6 +125,17 @@ def check_is_linear_program(program: Program) -> Optional[CheckFail]:
     return None
 
 
+def _has_variable(expr: Expr) -> bool:
+    if isinstance(expr, UnopExpr) and expr.operator == Unop.IVERSON:
+        return False
+    if isinstance(expr, VarExpr) and not expr.is_parameter:
```

Comment:
Shouldn't it also be possible to use constants here, since they are equivalent to nat / real / bool literals?

Reply:
Yes. But we usually just replaced constants directly after parsing so that they wouldn't occur during any further checks such as syntax/type checking. There might be more places that don't handle constants (the type checker?). It would certainly be cleaner to handle them everywhere though.

Reply:
Can we run with constant substitution in issues similar to this: 

```C++
#include <stdio.h>
#define MULTIPLY(a, b) a* b
int main()
{
    // The macro is expanded as 2 + 3 * 3 + 5, not as 5*8
    printf("%d", MULTIPLY(2 + 3, 3 + 5));
    return 0;
}
```
?

Reply:
Our constants don't have arguments, so that couldn't happen^^

Also, consider this program: `const x := y + y; z := x * x`. After constant substitution, we'd get an AST that represents the assignment `z := (y + y) * (y + y)` and _not_ `z := y + (y * y) + y`. This is because constant substitution is done on the AST, not on the textual level.

Reply:
> But we usually just replaced constants directly after parsing so that they wouldn't occur during any further checks such as syntax/type checking

I just checked, and the compiler only substitutes constants _after_ performing type checking. This should probably be changed. But I would also say that the typechecker should generally be able to handle constants, no?